cpp/include/cudf/detail/utilities/vector_factories.hpp

```diff
@@ -439,6 +441,40 @@ thrust::host_vector<typename Container::value_type> make_host_vector_sync(
   return make_host_vector_sync(device_span<typename Container::value_type const>{c}, stream);
 }
 
+/**
+ * @brief Asynchronously construct a pinned `cudf::detail::host_vector` of the given size
+ *
+ * @note This function does not synchronize `stream`.
+ *
+ * @tparam T The type of the vector data
+ * @param size The number of elements in the created vector
+ * @param stream The stream on which to allocate memory
+ * @return A host_vector of the given size
+ */
+template <typename T>
+host_vector<T> make_pinned_vector_async(size_t size, rmm::cuda_stream_view stream)
+{
+  return host_vector<T>(size, {cudf::get_pinned_memory_resource(), stream});
```

Comment:
I think this assumes that `thrust::host_vector` ctor does not synchronize. I _assume_ that's true but have you verified?

Reply:
We may be misusing the async allocator here. 
`thrust::host_vector` Has no notion of async allocation that I can find. It simply calls `allocate`.
We create `thrust::host_vector`s with `rmm_host_allocator`, which allocates as such:
```
  inline pointer allocate(size_type cnt)
  {
    if (cnt > this->max_size()) { throw std::bad_alloc(); }  // end if
    return static_cast<pointer>(
      mr.allocate_async(cnt * sizeof(value_type), rmm::RMM_DEFAULT_HOST_ALIGNMENT, stream));
  }
  ```
  i.e. a synchronous allocation ends up calling the `allocate_async`.
Most, if not all, thrust::host_vector constructors call `allocate` and then initialize all elements in the vector. Is this safe to do, given that we call `allocate_async` of the underlying allocator?

Reply:
OK, let's imagine you make a `host_vector` with an allocator that uses a managed pool MR. (It's host accessible, so this is conceivable). In that case what you get from `allocate_async` might still be in use in a kernel that accessed an allocation that was subsequently asynchronously deleted. So writing to it in any way that is not stream-ordered is not safe.  Initializing the elements of the vector on the host is not stream ordered. So I think you technically need to synchronize before initializing.  If the underlying allocator is not stream-ordered you don't need to synchronize.

Alternatively, could the initialization be done using [cudaLaunchHostFunction](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g05841eaa5f90f27124241baafb3e856f), which enqueues a host function call in a stream?

Reply:
I jumped ahead into implementing `host_uvector` that calls `allocate_async` and does not initialize the data. This fits well into the make_host_vector functions, where we get the data from a device_span anyway so the initial access to the host vector is stream-ordered.

Reply:
Seems like it should be in RMM though...

Reply:
Reduced the `host_uvector` to a TODO for a future PR.
Added sync to `rmm_host_allocator` to ensure `thrust::host_vector` init is safe.